packages/SwingSet/src/liveslots/watchedPromises.js

```diff
@@ -0,0 +1,162 @@
+// @ts-check
+import { assert } from '@agoric/assert';
+import { M } from '@agoric/store';
+import { parseVatSlot } from '../lib/parseVatSlots.js';
+
+export function makeWatchedPromiseManager(
+  syscall,
+  vrm,
+  vom,
+  cm,
+  convertValToSlot,
+  convertSlotToVal,
+) {
+  const { makeScalarBigMapStore } = cm;
+  const { defineDurableKind } = vom;
+
+  // watched promises by vpid: each entry is an array of watches on the
+  // corresponding vpid; each of these is in turn an array of a watcher object
+  // and the arguments associated with it by `watchPromise`.
+  let watchedPromiseTable;
+
+  function pseudoThen(p, vpid) {
+    function settle(value, wasFulfilled) {
+      const watches = watchedPromiseTable.get(vpid);
+      watchedPromiseTable.delete(vpid);
+      for (const watch of watches) {
+        const [watcher, ...args] = watch;
+        const handler = wasFulfilled ? watcher.onFulfilled : watcher.onRejected;
+        if (handler) {
+          handler(value, ...args);
+        }
+      }
+    }
+
+    p.then(
+      res => settle(res, true),
+      rej => settle(rej, false),
+    );
+  }
+
+  function loadWatchedPromiseTable() {
+    const deadPromisesRaw = syscall.vatstoreGet('deadPromises');
+    if (deadPromisesRaw) {
+      const deadPromises = new Set(deadPromisesRaw.split(','));
+      syscall.vatstoreDelete('deadPromises');
+      for (const [vpid, watches] of watchedPromiseTable.entries()) {
+        if (deadPromises.has(vpid)) {
+          watchedPromiseTable.delete(vpid);
+          for (const watch of watches) {
+            const [watcher, ...args] = watch;
+            if (watcher.onRejected) {
+              watcher.onRejected('vat upgraded', ...args);
+            }
```

Comment:
`onRejected` is user provided and can throw. Furthermore, user-code should not be executed synchronously in the middle of liveslots init.

```suggestion
            Promise.resolve().then(() => { watcher.onRejected('vat upgraded', ...args); });
```